lisa/tests/scheduler/load_tracking.py

```diff
@@ -971,7 +979,44 @@ def get_trace_cpu_util(self):
             phase_df = df_window(df, (start, end), method='pre', clip_window=True)
 
             for cpu in self.cpus:
-                util = phase_df[phase_df['cpu'] == cpu]['util']
+
+                cpu_phase_df = phase_df[phase_df['cpu'] == cpu].copy()
+
+                for task in tasks:
+                    if self.rtapp_profile[task].phases[phase-1].cpus.count(cpu):
+                        df_state_clipped = df_window(df_state[task], (start, end), method='pre', clip_window=True)
+                        cpu_phase_df = cpu_phase_df.join(df_state_clipped[[]], how='outer')
+
+                cpu_phase_df['cpu'].fillna(method='ffill', inplace=True)
+                cpu_phase_df = df_add_delta(cpu_phase_df)
+                cpu_phase_df['delta'] = cpu_phase_df['delta'].shift()
+                cpu_phase_df['prev_util'] = cpu_phase_df['util'].shift()
+
+                cpu_phase_df = cpu_phase_df[cpu_phase_df['cpu'].notna()]
+                cpu_phase_df = cpu_phase_df[cpu_phase_df['delta'].notna()]
+
+                cpu_capacity = self.plat_info['cpu-capacities']['rtapp'][cpu]
+
+                def compute_switch_phase_df(row):
+                    timestamp = row.name
+                    # Transform ftrace timestamp to PELT clock time.
+                    # Only CPU capacity (and not frequency) is considered
+                    # here. It has the bigger impact.
+                    delta = row.delta * cpu_capacity/PELT_SCALE
+                    last_update = timestamp - delta
+                    prev_util = row.prev_util
+                    window_shrink = 1e3
+                    activations = pd.Series([1, 1], index=[last_update, timestamp])
+                    simulated_phase_df = simulate_pelt(activations, init=prev_util,
+                                                       window=PELT_WINDOW/window_shrink,
+                                                       half_life=PELT_HALF_LIFE*window_shrink)
+                    util = simulated_phase_df.iloc[-1]
+                    return util
+
+                switch_loc = cpu_phase_df['util'].isnull()
+                cpu_phase_df.loc[switch_loc, 'util'] = cpu_phase_df.apply(compute_switch_phase_df, axis='columns')
```

Comment:
It looks like apply() should be called on `cpu_phase_df.loc[switch_loc]`, otherwise we do the (very costly) computation on the whole dataframe, and then throw away most of it.

Reply:
OK.

```
-                cpu_phase_df.loc[switch_loc, 'util'] = cpu_phase_df.apply(compute_switch_phase_df, axis='columns')
+                cpu_phase_df.loc[switch_loc, 'util'] = cpu_phase_df.loc[switch_loc].apply(compute_switch_phase_df, axis='columns')
```